doc/sla.md

```diff
@@ -0,0 +1,719 @@
+        serviceAgreement.setCondition(service, fingerprint, true);
+        return true;
+    }
+
+    // unsafe function!
+    function lockPayment(address receiver, bytes32 asset, uint256 amount, bytes32 condition) public payable returns(bytes32){
+        
+        require(msg.sender.balance > amount);
+        thisContract.transfer(amount);
+        // lock payments
+        Payment memory newPayment = Payment(true, amount, msg.sender, receiver, asset, condition);
+        bytes32 paymentId = keccak256(abi.encodePacked(amount, msg.sender, receiver, asset));
+        payments[paymentId] = newPayment;
+        return paymentId;
+    }
+
+    function requestAsset(bytes32 payment,
+                          bytes32 service) public isLocked(payment) isSender(payment) returns (bool){
+        // fulfil locked payment condition
+        fulfillCondition(service, this.requestAsset.selector);
+    }
+
+    function releasePayment(bytes32 payment, bytes32 service) public isLocked(payment) isReciever(payment) payable returns (bool){
+        // check that access was granted to the sender (consumer)
+        assert(serviceAgreement.getConditionStatus(payments[payment].condition) == true);
+        // TODO: transfer ether to reciever
+        payments[payment].receiver.transfer(payments[payment].amount);
+        // fulfill release payment condition
+        fulfillCondition(service, this.releasePayment.selector);
+        return true;
+    }
+
+
+}
+```
+
+### Event Consumer/Listener
+
+Ocean protocol is a dynamic network which means that it reacts based on the requests invoked by 
+the the network actors. Therefore, the event listener is meant to be the means to orchestrate 
+the event-driven approach. In order to use this approach, we have to define the following key points in the system:
+
+- **Event Creator:** in Ocean protocol, 
+events are generated by smart contract where they may or may not be associated with off-chain actions. The creator might be an actor, or 
+software invokes a transaction which in turn emits event.
+- **Event Manager**: It is a piece of software which acts as intermediary managing and processing events.
+When a manager receives a notification from a creator, it may pass it to event consumer directly or apply some 
+rules to process the event before passing it to the consumer. In ocean protocol, the network acts as a event manager.
+- **Event Consumer/Listener**: Consumer is an entity that needs to subscribe to an event manager (Ocean Network), and takes actions based on the event type/name.
+
+
+ 
+**Event listener Design**
+
+Event listener uses Ethereum [Web3](https://web3js.readthedocs.io/en/1.0/index.html) event watchers. The event listener catches the service 
+agreement setup event as follows:
+
+```
+event EventName(type<AgreementSetup>, service<SERVICEID>, status<BOOLEAN>)
+```
+For conditions events
+
+```
+event EventName(type<Condition>, service<SERVICEID>, codition<ConditionID>, status<BOOLEAN>)
```

Comment:
```Once a consumer "buys" the service, the service provider would like to signal that the service is fulfilled (automatically). How can this be implemented?```
This proposal states that the architecture will be based on ```Event-driven architecture``` and the components in ocean will implement this design , but the state of conditions and the whole agreement instance state will be stored on-chain in the storage contract. Therefore, even if you as a provider missed one the emitted events, you can call a ```view``` function on-chain that returns the current state.

Also, a condition has **three-state (1 --> true, 0 --> false, x --> unknown)** with a ```timeout```.

```Can a service provider that ignores events, effectively need/use service agreements at all? ```
- No, the system preserve the state condition on-chain and notify actors using events. So, events and conditions are 1st class citizens in the system.

```If not, can a service be provided, without service agreements?```
- No

Reply:
Thanks for the clarification.
From a perspective of getting new services on-board Ocean, it is absolutely imperative that existing services (especially those offered over HTTP/web services) have a low friction path to get on-board. 
These services are not going to listen to web3 events.

One way around this problem is to have a marketplace app (lets call it "Ocean service relay") that 

* allows a new (web) service (such as [openweather](https://openweathermap.org/api)) to register on Ocean
* uses the default service agreement template (and listens to web3 events on behalf of the web service).
* Manages service invocations (e.g. an Ocean user calls `squid.invokeService(serviceid, payload)`, therefore Ocean knows that the service has been called, and the results have been received.

The downside is that this relay may have to manage identity for that service. 

How feasible is the implementation of such a relay?

Reply:
I think this question would be out of the scope of the service agreement, but This will come later in a different document describing how to integrate existing marketplaces with ocean which I think as you pointed before, it would use one or two of Ocean's components especially squid and event listener.